### PR TITLE
[do not merge] upgrade and improve sphinx

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?

Our documentation dependencies are largely out of date and slow. I'd like to spend some time to improve this.

**Problem 1: Too slow**
There are a couple places for optimization:

1. parallelization with sphinx-build -j
2. cached writes (seems like most of the time is spent on writing output.
3. Updating sphinx. Seems like some problems were resolved in https://github.com/sphinx-doc/sphinx/issues/4576.

**Problem 2: Updating Sphinx and the Book Theme**

In this PR, I've updated the book theme and sphinx. IT provides a 10x speedup on repeated builds. However, this causes some of the sidebar titles to not be collapsible. We will need to fix this before merge.

~However, even with an updated sphinx, the build time has not changed.~
Current state:
- First build: make html  125.87s user 1.92s system 101% cpu 2:06.30 total
- Later build: make html  125.71s user 1.70s system 101% cpu 2:05.88 total

With an updated sphinx: 
 - First build: make html 100.91s user 1.47s system 101% cpu 1:40.82 total
 - Later builds: make html  11.95s user 0.78s system 115% cpu 11.060 total

With commenting out an extraneous file copy:
- Later builds: **make html  5.29s user 0.73s system 136% cpu 4.406 total**

TODO:
 - [ ] Fix sidebar title collapse (https://github.com/executablebooks/meta/discussions/597): Status = waiting for response from maintainers.

### CODE REVIEW

**Comment by richardliaw** [comment]
cc @maxpumperla

**Comment by maxpumperla** [comment]
@richardliaw thanks, this is great. I've compared doc build times with other projects, and it's actually not uncommon for sphinx builds to be much slower than ours (e.g. scikit takes 20+ minutes). like you said, writing outputs seems to be the bottleneck. so 11s seems very good.

What I noticed in our hackathon / POC is that e.g. in `mkdocs` most of the time is spent on API doc extraction and code inclusion. Plain markdown is blazing fast. If I remember correctly, `sphinx` was slower overall in parsing `rst` files. Maybe there's a compromise to be had, in that we have a flag that allows you to do a "quick build" of docs that excludes APIs and expensive examples. That might encourage more contributions and changes. just an idea.

**Comment by richardliaw** [comment]
OK, I found another place that triggers slowdown and pushed a patch. Current rebuild time is:

```
make html  6.74s user 0.75s system 128% cpu 5.828 total
```

**Comment by richardliaw** [comment]
OK, if we disable `sphinx.ext.viewcode`, we can get:
```
make html  5.29s user 0.73s system 136% cpu 4.406 total
```

Though I think I will not check this in, since every time you disable or enable an extension you trigger a full rebuild.

File: doc/source/conf.py [hunk 1]
```diff
@@ -180,21 +181,29 @@ def __getattr__(cls, name):
...
-    for f in glob.glob(os.path.join(source_dir, '*.rst')):
-        shutil.copy(f, gallery_dir)
-
-    # Copy inc files from source dir to gallery dir.
-    for f in glob.glob(os.path.join(source_dir, '*.inc')):
-        shutil.copy(f, gallery_dir)
+# If we do not set this environment variable, we will copy autogenerated
+# gallery files to a target directory. This creates a new file hash,
+# which then triggers rebuild.
+# If you set this environment variable, building can be faster,
```
Comment (maxpumperla) [inline doc/source/conf.py:187]: this should go into the README of the docs project.

**Comment by maxpumperla** [comment]
@richardliaw lack of theme customisability or convenience thereof is extremely annoying with the book theme ecosystem. The bump in version you propose instantly fixes our mobile view problems, but introduces this dropdown problem. Not sure how to go about this, since it's rather involved to customise this feature on our end.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Pull request closed without merging**